nemo/collections/asr/parts/jasper.py

```diff
@@ -309,21 +309,22 @@ def __init__(
                 nn.Linear(channels, channels // reduction_ratio, bias=False),
                 activation,
                 nn.Linear(channels // reduction_ratio, channels, bias=False),
-            )
+            ).float()
 
     def forward(self, x):
         # The use of negative indices on the transpose allow for expanded SqueezeExcite
         batch, channels, timesteps = x.size()[:3]
+        x = x.float()
```

Comment:
Do you need this .float()?

Reply:
Yes, definitely.

Otherwise, pooling will be done in fp16 and will cause instabilities.

Reply:
Same as above?

Reply:
That's the point of autocast. If it is stable in fp16, it will be autocasted. Otherwise it wouldn't be.
Are you claiming that our pooling layer isn't stable in fp16? If so, you need to open a bug on amp.

Reply:
`AvgPool` keeps the dtype of the input. But I see what you mean by stability. 

**What I observe is:** 
1. Train without any changes with AMP -> divergence. 
2. Train with just `x = x.float()` added  -> stable training. 

That's why I suspected that the pooling is the issue.

I've just dug deep into the pooling implementation. `AdaptiveAvgPool1d` is implemented via `AdaptiveAvgPool2d`. 
`AdaptiveAvgPool2d` has a short cut for kernels of size 1, calling `torch.mean` operation directly.

I didn't look further but my understanding is that `torch.mean` is implemented via float32 aggregation and then casting, so it should not cause any issues. 

So it's ever weirder that I see such behaviour. 

What would you recommend doing to pinpoint the exact issue?

Reply:
What happens if you move the pool inside the disable autocast context?

Reply:
Without float it fails due to precision mismatch.